activerecord/lib/active_record/reflection.rb

```diff
@@ -911,6 +911,14 @@ def association_class
           Associations::HasOneAssociation
         end
       end
+
+      def association_primary_key(klass = nil)
+        if primary_key = options[:primary_key]
+          @association_primary_key ||= -primary_key.to_s
+        else
+          primary_key(klass || self.klass)
+        end
+      end
```

Comment:
This and the test change in `activerecord/test/cases/reflection_test.rb` is somewhat of a revert of https://github.com/joshuay03/rails/commit/4ad0564245dc86a8e700ae2492c50fb9daf90967#diff-62fd84c8da9a63fa077a6fdea8acaac0df9e275b580aa1359a5c9416077fcd60, so I don't know if it's the correct patch to fix the `has_one` issue.

cc/ @kamipo Sorry for the tag, but I think you might be the best person to provide insight on this